### PR TITLE
Drawing: Use the state tree to get a mark's tool

### LINKED ISSUE #1363: Remove workflowSteps.tasks
**Package**
lib-classifier

**Is your feature request related to a problem? Please describe.**
#1307 removed `workflowSteps.tasks` then 19a73f1f hacked it back in in order to look up the tool that created a given mark. 

If we use `tool = getParent(mark)` to get the tool that created a mark, then we can simplify the drawing tool props and remove workflowSteps.tasks`.

### PR DESCRIPTION
**Pull request by eatyourgreens**
Use `getParentOfType` to get the tool that created a mark.
Remove redundant tool props from mark components.
Pass marks to annotations by reference.
Clear stale marks from a tool if the current annotation is empty.
Remove `workflowSteps.tasks`, which was only being used to look up the tool for a given mark.
Add tests for rendered marks in the subject viewer.

Package:
lib-classifier

Closes #1363.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Coverage increased with the new tests. That's good.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js [hunk 1]
```diff
@@ -38,18 +38,15 @@ const Drawing = types.model('Drawing', {
...
-      activeMarks.forEach(function addMarksToTools (mark) {
-        const newMark = clone(mark)
-        const tool = tryReference(() => self.tools[newMark.toolIndex])
-        tool && tool.marks.put(newMark)
-      })
-      self.updateAnnotation([])
+      // if the current annotation is empty, clear any leftover marks.
+      if (activeMarks.length === 0) {
+        self.tools.forEach(tool => tool.marks.clear())
+      }
```
Comment (srallen) [inline packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: Same comment about this change. A small explanation would be helpful since what this is doing is pretty different from before and I'm not sure how getting the mark's tool using MST's helper impacts this.

File: packages/lib-classifier/src/store/WorkflowStepStore.js [hunk 2]
```diff
@@ -60,10 +60,6 @@ const WorkflowStepStore = types
       }
 
       return false
-    },
-
-    get tasks () {
-      return Array.from(self.steps.values()).reduce((allTasks, step) => allTasks.concat(step.tasks), [])
```
[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/store/WorkflowStepStore.js]: Ok, I just noticed #1363. I should have more carefully read the PR description.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js [hunk 3]
```diff
@@ -19,34 +19,49 @@ describe('Components > DrawingToolMarks', function () {
...
   
-  it('should render two lines', function () {
-    const wrapper = shallow(<DrawingToolMarks tool={tool} svg={svg} />)
-    const marks = wrapper.find('Line')
-    expect(marks.length).to.equal(2)
+  it('should render a line', function () {
+    const wrapper = shallow(<DrawingToolMarks marks={marks} svg={svg} />)
+    const line = marks[0]
+    const mark = wrapper.find('Line')
+    expect(mark.prop('mark')).to.equal(line)
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js]: Good point. I've refactored the tests to clarify the names and remove `mark`.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js [hunk 4]
```diff
@@ -19,34 +19,49 @@ describe('Components > DrawingToolMarks', function () {
...
     })
 
     it('should render a delete button', function () {
-      const wrapper = shallow(<DrawingToolMarks activeMarkId='line1' tool={tool} svg={svg} />)
-      const line = tool.marks.get('line1')
+      const wrapper = shallow(<DrawingToolMarks activeMarkId='point1' marks={marks} svg={svg} />)
+      const point = marks[1]
       const deleteButton = wrapper.find('DeleteButton')
-      expect(deleteButton.prop('mark')).to.equal(line)
+      expect(deleteButton.prop('mark')).to.equal(point)
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js]: `.prop()` can only be called on a single node so I think this is ok. It'll fail if there's more than one delete button.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js [hunk 5]
```diff
@@ -38,18 +38,15 @@ const Drawing = types.model('Drawing', {
     }
 
     function complete () {
-      const newValue = self.marks.map(mark => detach(mark))
-      self.updateAnnotation(newValue)
+      self.updateAnnotation(self.marks)
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: Marks are stored on the annotation as `types.reference`, so there's no need to detach the mark from the tool any more.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: This is clearing any stale marks from each tool if the current annotation is empty when the task starts. The classifier doesn't seem to reset tasks on new subjects/classifications (as far as I can tell.) I think #1365 would help here too.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: There's a bit more discussion in https://github.com/zooniverse/front-end-monorepo/pull/1369#issuecomment-567427705. When I return to a drawing task, via the Back button, the task marks should still be present and editable.

Once we've decided how marks are submitted in the completed classification, then I think we can add code to clear out marks before the next classification starts.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js (hunk 1 above)
Comment (srallen) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: The `WorkflowStepStore` observes for the `onSubjectReady` action and resets itself which I think at one point managed the tasks and reset them, but some refactoring has happened since I wrote it, so I don't think it does that anymore. I think steps should be responsible for tasks and manage resetting itself rather than have this happen here. It's confusing to me.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js]: That's true, but that's implicitly relying on the Enzyme API which could be brittle.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js (hunk 1 above)
Comment (srallen) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: Making changes to the steps to manage clearing tasks is out of scope of this PR though.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/DrawingToolMarks/DrawingToolMarks.spec.js]: True.

File: packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: I found a bug in the subject/classification lifecycle, which should have been fixed in a recent PR (can't remember which one though.) New subjects rendered marks from the previous subject, because `onSubjectReady` was firing before the new classification was created.

Without docs, or examples, I do find it hard to follow what happens where in the classifier models. That's why I kept this code close to where it's needed. I'd be absolutely fine with new subjects calling something like `task.init()` or `task.reset()` on all workflow tasks before a new classification starts.

Perhaps rename `task.start()` to make it absolutely clear that it happens each time you navigate to that task during the current classification, rather than once when the new subject loads? The name seems ambiguous to me at the moment.

Comment (srallen) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: Yeah, a rename on `task.start()` sounds good. Do you mind opening a couple of issues to track these items?

Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: https://github.com/zooniverse/front-end-monorepo/pull/1339 fixed the bug where each new subject starts before `classificationComplete` for the previous subject.

Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/DrawingTask/models/DrawingTask.js]: Looking at the changes in #1339, the workflow steps store does reset each step task, by calling `task.start()`, when a new subject loads.
